controllers/hyperparametertuningjob/hyperparametertuningjob_controller.go

```diff
@@ -19,437 +19,292 @@ package hyperparametertuningjob
-
-		// Delete all spawned TrainingJobs, retry if failure.
-		err := ctx.HpoTrainingJobSpawner.DeleteSpawnedTrainingJobs(ctx, ctx.Job)
-		if err != nil {
-			log.Info("Not all associated TrainingJobs jobs were deleted, will retry", "error", err)
-			return RequeueAfterInterval(r.PollInterval, nil)
+		if err = r.createHyperParameterTuningJob(ctx); err != nil {
+			return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to create hyperparameter tuning job"))
 		}
 
-		return r.removeFinalizerAndUpdate(ctx)
-	default:
-		log.Info("Job is in unknown status")
-		return NoRequeue()
+		if ctx.TuningJobDescription, err = ctx.SageMakerClient.DescribeHyperParameterTuningJob(ctx, ctx.TuningJobName); err != nil {
+			return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to describe SageMaker hyperparameter tuning job"))
+		}
 	}
-}
 
-func (r *HyperparameterTuningJobReconciler) addFinalizerAndRequeue(ctx reconcileRequestContext) (ctrl.Result, error) {
-	ctx.Job.ObjectMeta.Finalizers = append(ctx.Job.ObjectMeta.Finalizers, SageMakerResourceFinalizerName)
-	ctx.Log.Info("Add finalizer and Requeue")
-	prevGeneration := ctx.Job.ObjectMeta.GetGeneration()
-	if err := r.Update(ctx, &ctx.Job); err != nil {
-		ctx.Log.Error(err, "Failed to add finalizer", "StatusUpdateError", ctx)
-		return RequeueIfError(err)
-	}
+	switch ctx.TuningJobDescription.HyperParameterTuningJobStatus {
+	case sagemaker.HyperParameterTuningJobStatusInProgress:
+		ctx.HPOTrainingJobSpawner.SpawnMissingTrainingJobs(ctx, *ctx.TuningJob)
 
-	return RequeueImmediatelyUnlessGenerationChanged(prevGeneration, ctx.Job.ObjectMeta.GetGeneration())
-}
+		if controllers.HasDeletionTimestamp(ctx.TuningJob.ObjectMeta) {
+			// Request to stop the job
+			if _, err := ctx.SageMakerClient.StopHyperParameterTuningJob(ctx, ctx.TuningJobName); err != nil && !clientwrapper.IsStopHyperParameterTuningJob404Error(err) {
+				return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to delete hyperparameter tuning job"))
+			}
+			// Describe the new state of the job
+			if ctx.TuningJobDescription, err = ctx.SageMakerClient.DescribeHyperParameterTuningJob(ctx, ctx.TuningJobName); err != nil {
+				return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to describe SageMaker hyperparameter tuning job"))
+			}
+		}
+		break
 
-func (r *HyperparameterTuningJobReconciler) getHyperParameterTuningJobName(state hpojobv1.HyperparameterTuningJob) string {
-	return GetGeneratedJobName(state.ObjectMeta.GetUID(), state.ObjectMeta.GetName(), 32)
-}
+	case sagemaker.HyperParameterTuningJobStatusCompleted:
+		if err = r.addBestTrainingJobToStatus(ctx); err != nil {
+			return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to add best training job to status"))
+		}
```

Comment:
This should happen for all statuses. The best training job can even be the first training job to complete -- it is updated over the lifetime of the tuning job.